src/profile-logic/flame-graph.js

```diff
@@ -38,229 +40,255 @@ export type FlameGraphTiming = Array<{
+  //  - callNodesAtDepth[k][pendingRangeStartAtDepth[k]..] are "pending"
+  //
+  // A node starts out as "pending" when we initially add it to the row.
+  // A node becomes "finished" once we've decided to process its children.
+  //
+  // This is used to queue up a bunch of siblings before we process their
+  // children.
+  // We need to queue up nodes before we can process their children because
+  // we can only process children once their parents are in the right order.
+  const callNodesAtDepth = [[]];
+  const pendingRangeStartAtDepth = [0];
 
-    if (totalTime[callNodeIndex] === 0) {
-      continue;
-    }
+  // At the beginning of each turn of this loop, add currentCallNode and all its
+  // siblings as "pending" to row[currentDepth], ordered by name. Then find the
+  // first pending call node with children, and go to the next iteration.
+  let currentCallNode = 0;
+  let currentDepth = 0; // always set to depth[currentCallNode]
+  outer: while (true) {
+    // assert(depth[currentCallNode] === currentDepth);
 
-    const parent = callNodeTable.prefix[callNodeIndex];
-    if (parent === -1) {
-      roots.push(callNodeIndex);
-      continue;
+    // Add currentCallNode and all its siblings to the current row. Ensure correct
+    // ordering when inserting each sibling.
+    const rowAtThisDepth = callNodesAtDepth[currentDepth];
+    const siblingIndexRangeStart = rowAtThisDepth.length; // index into rowAtThisDepth
+    for (
+      let currentSibling = currentCallNode;
+      currentSibling !== -1;
+      currentSibling = nextSibling[currentSibling]
+    ) {
+      const siblingIndexRangeEnd = rowAtThisDepth.length;
+      if (siblingIndexRangeStart === siblingIndexRangeEnd) {
+        // This is the first sibling that we see. We don't need to compute an
+        // insertion index because we don't have any other siblings to compare
+        // to yet.
+        rowAtThisDepth.push(currentSibling);
+      } else {
+        // There are other siblings already present in rowAtThisDepth[siblingIndexRangeStart..].
+        // Do an ordered insert, to keep siblings ordered by function name.
+        // assert(siblingIndexRangeStart < siblingIndexRangeEnd)
+        const thisFunc = func[currentSibling];
+        const funcName = stringTable.getString(funcTableNameColumn[thisFunc]);
+        const insertionIndex = bisectionLeftByStrKey(
+          rowAtThisDepth,
+          funcName,
+          (cn) => stringTable.getString(funcTableNameColumn[func[cn]]),
+          siblingIndexRangeStart,
+          siblingIndexRangeEnd
+        );
+        rowAtThisDepth.splice(insertionIndex, 0, currentSibling);
+      }
     }
 
-    const funcName = thread.stringTable.getString(
-      thread.funcTable.name[callNodeTable.func[callNodeIndex]]
-    );
+    // Now currentCallNode and all its siblings have been added to the row, and
+    // they are ordered correctly. Find a queued up node in this row which has
+    // children, and descend into it.
```

Comment:
Sometimes we "descend" and sometimes we "go up", it might be useful to be more explicit about the various cases. Especially that the only goal for the rest of this loop is to find the next value for `currentCallNode`. That would make it easier to follow. Maybe it would be possible to extract this part in a function to make it easier to follow (but maybe not).

Reply:
Would it help if I replaced "descend" with "go down" or "go deeper"?

I'll try to make the comments a bit clearer.

I'm not sure if anything can be easily pulled out into a separate function; the algorithm would be easier to follow if it used recursion, but then it would use recursion... and I'm a bit afraid of the stack usage for profiles with deep stacks, and of the function call overhead.